### PR TITLE
Replace behave framework with pytest-bdd and add DataSetComparator

### PR DESCRIPTION
**Pull request by yixinglu**
Depends on nebula-python ValueWrapper compare implementation.

### CODE REVIEW

File: tests/tck/conftest.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+from tests.tck.utils.nbv import parse
+
+
+def _table(text):
+    lines = text.splitlines()
+    assert len(lines) >= 1
+
+    def parse_line(line):
+        return list(map(lambda x: x.strip(),
+                        filter(lambda x: x, line.split('|'))))
```
Comment (dutor) [inline tests/tck/conftest.py:19]: This will filter out the empty cells, right? Is that as expected?

File: tests/tck/utils/comparator.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        if self._order:
+            return all(self.compare_row(l, r) for (l, r) in zip(lhs, rhs))
+        return self._compare_list(lhs, rhs, self.compare_row)
+
+    def compare_value(self, lhs: ValueWrapper, rhs: ValueWrapper):
+        '''
+        lhs and rhs represent response data and expected data respectively
+        '''
+        if lhs.is_null():
+            return rhs.is_null()
```
Comment (dutor) [inline tests/tck/utils/comparator.py:40]: Shall we elaborate the individual null types?

**Review by dutor — COMMENTED** [review]
Excellent

File: tests/tck/conftest.py (hunk 1 above)
Comment (yixinglu) [reply tests/tck/conftest.py]: yeah. We don't allow following case `| nothing here |` in expected table, so is safe to filter the empty cells.

File: tests/tck/utils/comparator.py (hunk 2 above)
Comment (yixinglu) [reply tests/tck/utils/comparator.py:40]: Maybe. But we can let it go at present. When we encounter those test cases, it's not late to fix it. Will add some todo comments.

File: ci/test.sh [hunk 3]
```diff
@@ -91,16 +91,12 @@ function run_ctest() {
...
-        -n=8 \
+    pytest -n 8 --build_dir=$BUILD_DIR \
         --dist=loadfile \
         --debug_log=false \
-        ${@:1} \
-        $testpath
+        ${@:1}
 
-    $BUILD_DIR/tests/ntr --debug_log=false ${@:1} $PROJ_DIR/tests/job/*
+    # $BUILD_DIR/tests/ntr --debug_log=false ${@:1} $PROJ_DIR/tests/job/*
```
Comment (laura-ding) [inline ci/test.sh:99]: Why comment it out

File: tests/tck/utils/comparator.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            and self.compare_node(l.end_node, r.end_node)
+            and self.compare_edge(l.relationship, r.relationship)
+            for (l, r) in zip(lhs, rhs))
+
+    def compare_edge(self, lhs: Relationship, rhs: Relationship):
+        if not lhs == rhs:
+            return False
+        if not self._strict:
+            return True
+        return self.compare_map(lhs.propertys(), rhs.propertys())
```
Comment (laura-ding) [inline tests/tck/utils/comparator.py:95]: need to compare edge key

File: tests/tck/utils/comparator.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    def compare_node(self, lhs: Node, rhs: Node):
+        if lhs.get_id() != rhs.get_id():
+            return False
+        if not self._strict:
+            return True
+        for tag in lhs.tags():
```
Comment (laura-ding) [inline tests/tck/utils/comparator.py:102]: check the size of the tags firstly

File: tests/tck/utils/comparator.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        if lhs.is_edge():
+            return rhs.is_edge() and self.compare_edge(lhs.as_relationship(),
+                                                       rhs.as_relationship())
+        if lhs.is_path():
+            return rhs.is_path() and self.compare_path(lhs.as_path(),
+                                                       rhs.as_path())
+        return False
+
+    def compare_path(self, lhs: PathWrapper, rhs: PathWrapper):
+        return all(
```
Comment (laura-ding) [inline tests/tck/utils/comparator.py:84]: check the path size firstly

File: tests/tck/utils/comparator.py (hunk 4 above)
Comment (yixinglu) [reply tests/tck/utils/comparator.py:95]: nebula python client has compared edge keys in `Relationship::__eq__`. `lhs == rhs` will call the equal function.

File: tests/tck/utils/comparator.py (hunk 5 above)
Comment (yixinglu) [reply tests/tck/utils/comparator.py:102]: fixed it.

File: tests/tck/utils/comparator.py (hunk 6 above)
Comment (yixinglu) [reply tests/tck/utils/comparator.py:84]: fixed it.

File: ci/test.sh (hunk 3 above)
Comment (yixinglu) [reply ci/test.sh:99]: will fixed job test cases in next PR

Comment (yixinglu) [reply ci/test.sh:99]: will fix job test cases in next PR

**Review by dutor — APPROVED** [review]

File: tests/conftest.py [hunk 7]
```diff
@@ -62,6 +73,184 @@ def pytest_configure(config):
...
+            load_csv_data(pytestconfig, conn_pool, "student")
+            fn.write_text("student")
+            logging.info(f"session-{worker_id} load student csv data")
+            load = True
+        else:
+            logging.info(
+                f"session-{worker_id} need not to load student csv data")
+    yield
+    if load:
+        os.remove(str(fn))
```
Comment (laura-ding) [inline tests/conftest.py:216]: you can define one function, and make the space_name as input args

File: tests/query/v1/test_find_path.py [hunk 8]
```diff
@@ -589,7 +589,7 @@ def test_path_limit(self):
             ]
         }
         self.check_column_names(resp, expected_data["column_names"])
-        self.check_path_result_without_prop(resp.rows(), expected_data["rows"])
+        # self.check_path_result_without_prop(resp.rows(), expected_data["rows"])
```
Comment (laura-ding) [inline tests/query/v1/test_find_path.py:592]: why comment?

**Review by laura-ding — APPROVED** [review]
excellent! please modify the readme for the tests

File: tests/query/v1/test_find_path.py (hunk 8 above)
Comment (yixinglu) [reply tests/query/v1/test_find_path.py:592]: will be fixed in next PR.

File: tests/conftest.py (hunk 7 above)
Comment (yixinglu) [reply tests/conftest.py:216]: make sense, I am thinking how to optimize these fixtures if the `yield`sentence in different functions.

**Pull request merged**